sql/src/main/java/io/crate/execution/engine/distribution/merge/BatchPagingIterator.java

```diff
@@ -166,14 +154,7 @@ private void raiseIfClosedOrKilled() {
 
     @Override
     public void kill(@Nonnull Throwable throwable) {
-        CompletableFuture<? extends Iterable<? extends KeyIterable<Key, Row>>> loading;
-        synchronized (this) {
-            killed = throwable;
-            loading = this.currentlyLoading;
-        }
+        killed = throwable;
         close();
-        if (loading != null) {
-            loading.completeExceptionally(throwable);
```

Comment:
Would it have worked if we instead called `loading.cancel(true)` here?

I think generally if a component invokes some other action, and decides to prematurely cancel that action, it should be allowed to do so. I agree that it shouldn't release resources created by the other component, but telling this component to stop doing what it was doing should be fine?

Reply:
Not sure where you heading to, `loading.cancel(true)` would also complete the future and thus release it too early. I think only the component which *owns*  the future should complete it as it may need to stop/cleanup something upfront. Otherwise the component would have to register its cleanup/stop//release work on the future completion.

Can you please elaborate a bit more what you have in mind?

Reply:
We currently have the pattern that we 

a) Kill the batchIterator, if the operation already started
b) fail the consumer if the operation hasn't started.


In the first case, the idea is that the consumer will eventually receive a failure (due to moveNext or another method raising an exception).

So we've a pipeline:

    rootBatchIterator                  -*
      transformationA-batchIterator      \ propagates to the inner most instance
        transformationB-batchIterator  <-- kill invoked here
          rowconsumer

(A and B are for example `ForwardingBatchIterator` sub-classes)


The pipeline is a more complex in the `BatchPagingIterator` case, but I think the general idea is the same: The kill should ideally short-circuit whatever action is encapsulated via `loadNextBatch`. If not, it probably means that the operation has to finish running (and once it finished, we discard the result. (Or maybe it is already killed via other means, but I don't think it's good to rely on that)).

So my thoughts are in the direction of:

How can we interrupt/short-circuit the action triggered via `loadNextBatch` in a way that it still gives the action the option to handle the interrupt. Calling `completeExceptionally` is by-passing that, but maybe `fetchMore` returns a future that has a custom `cancel` implementation, or we return a different result from `fetchMore` that exposes a `Killable` property itself?


I think it's okay to go ahead with this fix now, but I think the concept of interrupting the action encapsulated via `loadNextBatch` is legitimate and nice to have to react faster to kill requeusts.

Reply:
Seems like I have a different understanding of our patterns as I thought we only propagate kill calls *down* but not *up* to avoid such situations (expecting all errors are forwarded and are eventually ending up on the consumer).
The kill calls are propagated via our task implementations which will then kill all their tied components, so we can rely on that upper components in that flow are already killed.
Afaik it is like that right now and is not a *maybe killed via other means*.

Imho allowing/supporting to kill upper components will make this more complex and I'm not sure how we benefit from it.

Reply:
I guess if it is up or down is a matter of perspective.

Let me try to explain myself in a different way.

Currently we kill at the task level, in the `CollectTask` this looks like

```java
@Override
public void innerKill(@Nonnull Throwable throwable) {
    if (batchIterator != null) {
        batchIterator.kill(throwable);
    }
    innerClose();
}
```


Here `batchIterator` is the outer-most BatchIterator. Each projection/projector wraps a inner projection.

So we might have: `filter(group(luceneBatchIterator))`.

So

```
  filter.kill(..)
    invokes
      group.kill(...)
```


GROUP BY is implemented as a pipeline breaker (ignoring the optimized cases for now), and all the expensive grouping logic is encapsulated via `loadNextBatch`. We don't want to wait for the `loadNextBatch` action to finish, but we want to abort it eagerly.

In the group case `loadNextBatch` fully consumes the inner BatchIterator, so we can delegate `kill` to that. In `CollectingBatchIterator`:

```
public static <T, A> BatchIterator<T> newInstance(BatchIterator<T> source,
                                                  Collector<T, A, ? extends Iterable<? extends T>> collector) {
    return newInstance(
        source::close,
        source::kill,
        () -> BatchIterators.collect(source, collector),
        source.involvesIO()
    );
}
```

So we have:


```
  filter.kill(..)
    invokes
      group.kill(...)
        invokes
          source.kill
```

`luceneBatchIterator` at that point raises a exception on `moveNext`, which in turns causes the `loadNextBatch` of the group operation to exit early.



I think this kind of "kill connection" is not present in the `BatchPagingIterator`, and I'm wondering if it should be there as well.

Reply:
> I think this kind of "kill connection" is not present in the BatchPagingIterator, and I'm wondering if it should be there as well.

Given that we're requesting more data from the upstream I think there is little that we can do to abort that, so I guess the change is fine. 

But I wanted to make sure we're on the same page that it is generally not a concept issue to interrupt actions that a batchIterator triggered

Reply:
I think its worth aligning our visions using different channels.